src/uct/tcp/tcp_iface.c

```diff
@@ -124,32 +123,44 @@ static ucs_status_t uct_tcp_iface_event_fd_get(uct_iface_h tl_iface, int *fd_p)
 unsigned uct_tcp_iface_progress(uct_iface_h tl_iface)
 {
     uct_tcp_iface_t *iface = ucs_derived_of(tl_iface, uct_tcp_iface_t);
+    unsigned read_events   = 0;
+    unsigned count         = 0;
     struct epoll_event events[UCT_TCP_MAX_EVENTS];
     uct_tcp_ep_t *ep;
-    unsigned count;
-    int i, nevents;
-    int max_events;
-
-    ucs_trace_poll("iface=%p", iface);
-
-    max_events = ucs_min(UCT_TCP_MAX_EVENTS, iface->config.max_poll);
-    nevents = epoll_wait(iface->epfd, events, max_events, 0);
-    if ((nevents < 0) && (errno != EINTR)) {
-        ucs_error("epoll_wait(epfd=%d max=%d) failed: %m", iface->epfd,
-                  max_events);
-        return 0;
-    }
-
-    count = 0;
-    for (i = 0; i < nevents; ++i) {
-        ep = events[i].data.ptr;
-        if (events[i].events & EPOLLIN) {
-            count += ep->rx.progress(ep);
+    int i, nevents, max_events;
+
+    do {
+        max_events = ucs_min(iface->config.max_poll - read_events,
+                             UCT_TCP_MAX_EVENTS);
+
+        nevents = epoll_wait(iface->epfd, events, max_events, 0);
+        if (ucs_unlikely((nevents < 0))) {
+            if (errno == EINTR) {
+                /* force a new loop iteration */
+                nevents = 1;
```

Comment:
why you set nevents here? this var is updated on line 136

Reply:
ops, this needs to be set to `max_event` to ensure the new iteration, because we can do try again to handle interrupted during call case